enhancements/cluster-logging/forward_to_cloudwatch.md

```diff
@@ -0,0 +1,249 @@
+see-also:
+superseded-by:
+---
+
+# Forward to Cloudwatch
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+[Amazon Cloudwatch][aws-cw] is a hosted monitoring and log storage service.
+This proposal extends the `ClusterLogForwarder` API with an output type for Cloudwatch.
+
+## Motivation
+
+Amazon Cloudwatch is a popular log store.
+We have requests from external and  Red Hat-internal customers to support it.
+
+### Goals
+
+Enable log forwarding to Cloudwatch.
+
+### Non-Goals
+
+Enable Cloudwatch metric collection.
+
+## Proposal
+
+### Cloudwatch streams and groups
+
+[Cloudwatch][concepts] defines *log groups* and *log streams*. To paraphrase the documentation:
+
+>  A log stream is a sequence of log events that share the same source ... For example, an Apache access log on a specific host.
+
+>  Log groups define groups of log streams that share the same retention, monitoring, and access control settings ... For example, if you have a separate log stream for the Apache access logs from each host, you could group those log streams into a single log group called MyWebsite.com/Apache/access_log.
+
+In other words a *log stream* corresponds to the smallest log-producing units.
+*Log groups* are used to group larger sets of related logs together.
+
+We must consider *container logs*, and *node logs*.
+- "application" logs are always container logs
+- "infrastructure" logs are a mix of container and node logs
+- "audit" logs are all node logs.
+
+For *container-scoped* logs we auto-create a log stream for *each container*, using the container UUID as the log stream name.
+- The log stream name is just a unique identifier for the source container.
+- Log *entries* include meta-data (container name, namespace etc.) for the user to search/index.
+  The log stream name is *not* needed for indexing or searching logs.
+- Log streams are for a *single source*, we cannot send logs from multiple nodes on the same log stream.
+
+For *node-scoped* logs we will create "audit"" and "infrastructure" streams for each node: e.g. `node-id/audit`.
+
+By default the forwarder creates 3 *log groups* for each of the log categories:  "application", "infrastructure" and "audit".
+Application logs can also be grouped by namespace, k8s labels, or forwarder configuration, see below.
+
+### API fields
+
+New API fields in the `output.cloudwatch` section:
+
+- `region`: (string) AWS region name, required to connect.
+- `retentionDays`: (number) Number of days to keep logs. Defaults to cloudwatch default retention.
+- `groupByNamespace`: (bool) If true, set the Cloudwatch group name to the source namespace.
```

Comment:
Keep in mind, one of the problems with Kubernetes "namespace" names is that they can be re-used over time.  So if somebody defines a very long retention period in Cloudwatch, but namespaces change, they might have unwanted streams landing in the same group.

Consider using namespace UUID instead.  See also https://github.com/kubernetes/kubernetes/issues/58638.

Reply:
My current assumption is that if you create a forwarder configuration with "namespaces: [foo]",
you want logs from any namespace named "foo", present or future. Presumably the name "foo" means something to you, and if you delete and re-create it, then it is still "the same foo" in some sense. This seems to also be the interpretation of tools like oc, for example: 
    oc get -ns foo cm --watch
will keep printing out configmaps created in namespace 'foo' even if I delete and re-create it.

There may be use cases for supporting namespace UUIDs *as well as* names, but surely not *instead* of? Can you point me to existing k8s APIs/tools that use namespace UUIDs instead of names so I can get an idea of what the use cases are?

Reply:
Unfortunately, it is our responsibility to properly communicate that to the users, but this is an "action-at-a-distance" problem.  The retention times can be so long, that one namespace name can go away, and then be recreated by somebody else in charge.  If there is any sensitive information in those logs collected, we are not helping them keep them separate.

Using namespace UUIDs is the only way to be sure.

> My current assumption is that if you create a forwarder configuration with "namespaces: [foo]", you want logs from any namespace named `foo"`, present or future.  Presumably the name `foo` means something to you, and if you delete and re-create it, then it is still "the same `foo`" in some sense.

With short retention times it is very like the "you" is the same person or group.  With long retention times, it is much more likely the "you" changes, and the presumptions we make will have changed.

> This seems to also be the interpretation of tools like `oc`, for example: `oc get -ns foo cm --watch` will keep printing out `configmap`s created in namespace '`foo` even if I delete and re-create it.

True, but there is no history for namespaces in Kubernetes (which is also a problem, the Kube community has not been around long enough to realize the issue yet).  So there is no facility for Kubernetes to do anything different.

But that short retention period does not mean it translates to long term log retention.

It is a design flaw of Kubernetes that operational semantics use name lookups instead of UUIDs.  All of Kubernetes would operate the same using UUIDs, but not the other way around.

> There may be use cases for supporting namespace UUIDs _as well as_ names, but surely not _instead_ of?

You can all solve all use cases with namespace UUIDs, you can't handle all use cases with namespace names because names are reused by Kubernetes.

> Can you point me to existing k8s APIs/tools that use namespace UUIDs instead of names so I can get an idea of what the use cases are?

As stated above, log retention times are much, much longer than operational Kubernetes, so they don't equate or translate.  We have to be thinking about this in our solution to help users make sure they don't expose sensitive information inadvertently.

Reply:
I don't understand how you want to use UUIDs. Do you want to 
a) use them in forwarder configuration instead of namespace names? 
b) have the forwarder look up the UUID based on a configured namespace name?

a) makes configuration of the forwarder tricky but I guess if you want to use UUIDs in place of namespace names that's inevitable. 

b) raises the question of *when* the forwarder should look up the UUID. Given the eventually-consistent nature of k8s, we can't assume that the target namespace will exist when the forwarder is started, and even if it does we don't know if it is the _intended incarnation_ of that namespace.

I would note again that the namespace UUID is part of the meta-data we store, so it is possible after the fact to determine which incarnation of a namespace a given log came from.

I'm open to supporting UUIDs but given that using names is such a widespread practice (even if it is a design flaw) I think we would have to support both.

Reply:
> I don't understand how you want to use UUIDs. Do you want to
> a) use them in forwarder configuration instead of namespace names?

Yes, the target log group would be by namespace UUID instead of namespace name.

> b) have the forwarder look up the UUID based on a configured namespace name?

As you wrote below, this is already happening for the most part with the OpenShift fluentd configuration today.

> a) makes configuration of the forwarder tricky but I guess if you want to use UUIDs in place of namespace names that's inevitable.

You already have that today.

> b) raises the question of _when_ the forwarder should look up the UUID. Given the eventually-consistent nature of k8s, we can't assume that the target namespace will exist when the forwarder is started, and even if it does we don't know if it is the _intended incarnation_ of that namespace.

Can you explain more of what that sentence above means?

The whole purpose of https://github.com/kubernetes/kubernetes/issues/58638 is to make it so that there is no need for a namespace UUID lookup by name when collecting logs: when the container log file is created, it is created with the namespace UUID and the namespace name so that no lookup is required, and the logs can be safely forwarded to another system where the UUID can be used to correctly associate logs from different containers in their namespace.

> I would note again that the namespace UUID is part of the meta-data we store, so it is possible after the fact to determine which incarnation of a namespace a given log entry came from, and to filter on UUID if you need to focus on one incarnation.

The problem is that if you don't use UUID to store the logs remotely in the Log Group, you can't help the user ensure that logs from one namespace are not made visible to the wrong users who have access to a Log Group.

> I'm open to supporting UUIDs but given that using names is such a widespread practice (even if it is a design flaw) I think we would have to support both.

Sure, supporting both is possible, as long as we explain to the users the potential issues with both and how they can avoid them, and how the log forwarder can help or hinder the proper log file association.

Reply:
The current proposal for cloudwatch is to use the container UUID to create log *streams*, which is as unique as using a namespace UUID. 
Log *groups* are created at the user's discretion via configuration of the forwarder. Currently we allow log type (app/infra/auth), namespace *name* and (shortly) POD label selectors as criteria. 
We can easily add namespace UUID as an additional input selection criterion. However if we need to go to that level of detail, perhaps we should add general field selectors rather than treating UUID as a special case, since there might be other fields of interest. 
What do you think? There are 2 separate concerns here: 
1. user requirements/UI design
2. implementation
I realize UUID has special case status in k8s but we should consider 1. as well as 2. to decide between UUID-as-speical-case vs. general-field-selector UIs.

Reply:
> The current proposal for cloudwatch is to use the container UUID to create log _streams_, which is as unique as using a namespace UUID.

If UUIDs are worth their salt then no container from any node across all possible clusters that could ever exist in our lifetime would end up with the same UUID.  =)

So this proposal does not give the user another way to choose how to create log streams, which is smart.  Too much flexibility in the first implementation leads to complexity, making it much harder to verify.

> Log _groups_ are created at the user's discretion via configuration of the forwarder. Currently we allow log type (app/infra/auth), namespace _name_ and (shortly) POD label selectors as criteria.

Here it might be prudent and avoid providing too many ways to create log groups on the first pass, so adding POD label selectors might be worth waiting on to avoid testing complexity.

There is another case that we have to consider here as well: multiple clusters sending to one CloudWatch instance.  This is likely to be a common use-case given the Kubernetes clusters tend to be small.  If this ends up being true, we'll want to make sure the user is okay with mixing logs from different clusters in the same groups, or adding a way to create groups unique to a cluster.  If shared CloudWatch instances ends up not being true, no real issue.

Is CloudWatch query access is by log group?  If so, for the "log type" and "namespace name" case, clusters sharing the same CloudWatch instance would potentially have a problem limiting access.  Using "namespace UUID" would not be a problem, because the namespaces are unique across the clusters.

> We can easily add namespace UUID as an additional input selection criterion. However if we need to go to that level of detail, perhaps we should add general field selectors rather than treating UUID as a special case, since there might be other fields of interest.
> What do you think? There are 2 separate concerns here:
> 
> 1. user requirements/API design
> 2. implementation details/performance
> 
> I realize UUID has special case status in k8s but we should consider 1. as well as 2. to decide between UUID-as-special-case vs. general-field-selector API.

It seems easier to add support for namespace UUID before the POD label selector or a general field selector.  The problem we have is when a log is collected and ends up not having the particular field to use, what then?  If we ensure a log always has a namespace UUID, namespace name, and log type, then we can have a well behaved system.  If we can't ensure those fields are always present, then we have to define what happens in those cases.

Reply:
Thanks for your patience, that was a worthwhile discussion.
I think the new proposal is exactly what you want, and I have resolved my confusion caused by mixing new cloudwatch features with how existing "generic" forwarder features can be used with cloudwatch. The two are now fully orthogonal which is as it should be.